File: src/main/java/net/Chidoziealways/everythingjapanese/screen/ModMenuTypes.java
package net.Chidoziealways.everythingjapanese.screen;

import net.Chidoziealways.everythingjapanese.EverythingJapanese;
import net.Chidoziealways.everythingjapanese.screen.custom.growthchamber.GrowthChamberMenu;
import net.Chidoziealways.everythingjapanese.screen.custom.pedestal.PedestalMenu;
import net.minecraft.core.registries.Registries;
import net.minecraft.world.inventory.MenuType;
import net.minecraftforge.common.extensions.IForgeMenuType;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.RegistryObject;

public class ModMenuTypes {
    public static final DeferredRegister<MenuType<?>> MENUS =
            DeferredRegister.create(Registries.MENU, EverythingJapanese.MOD_ID);

    public static final RegistryObject<MenuType<PedestalMenu>> PEDESTAL_MENU =
            MENUS.register("pedestal_menu", () -> IForgeMenuType.create(PedestalMenu::new));

    public static final RegistryObject<MenuType<GrowthChamberMenu>> GROWTH_CHAMBER_MENU =
            MENUS.register("growth_chamber_menu", () -> IForgeMenuType.create(GrowthChamberMenu::new));

    public static void register(IEventBus eventBus) {
        MENUS.register(eventBus);
    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/screen/ModMenuTypes.kt
package net.Chidoziealways.everythingjapanese.screen

import net.Chidoziealways.everythingjapanese.JAPANESE_MOD_ID
import net.Chidoziealways.everythingjapanese.screen.custom.calligraphytable.CalligraphyTableMenu
import net.Chidoziealways.everythingjapanese.screen.custom.growthchamber.GrowthChamberMenu
import net.Chidoziealways.everythingjapanese.screen.custom.pedestal.PedestalMenu
import net.minecraft.core.registries.Registries
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.world.entity.player.Inventory
import net.neoforged.bus.api.IEventBus
import net.neoforged.neoforge.common.extensions.IMenuTypeExtension
import net.neoforged.neoforge.registries.DeferredRegister
import thedarkcolour.kotlinforforge.neoforge.forge.getValue
import java.util.function.Supplier

object ModMenuTypes {
    val MENUS = DeferredRegister.create(Registries.MENU, JAPANESE_MOD_ID)

    val PEDESTAL_MENU by MENUS.register(
        "pedestal_menu",
        Supplier {
            IMenuTypeExtension.create { pContainerId: Int, inv: Inventory, extraData: FriendlyByteBuf ->
                PedestalMenu(
                    pContainerId,
                    inv,
                    extraData
                )
            }
        })

    val CALLIGRAPHY_TABLE_MENU by MENUS.register(
        "calligraphy_table_menu") { ->
        IMenuTypeExtension.create { containerId, inv, _ ->
            CalligraphyTableMenu(
                containerId,
                inv
            )
        }
    }

    val GROWTH_CHAMBER_MENU by
        MENUS.register(
            "growth_chamber_menu",
            Supplier {
                IMenuTypeExtension.create<GrowthChamberMenu?> { pContainerId: Int, inv: Inventory, extraData: FriendlyByteBuf ->
                    GrowthChamberMenu(
                        pContainerId,
                        inv,
                        extraData
                    )
                }
            })

    fun register(eventBus: IEventBus) {
        MENUS.register(eventBus)
    }
}
